lib/core/models/repositories/roleRepository.js

```diff
@@ -181,25 +195,118 @@ class RoleRepository extends Repository {
           errorsManager.throw('security', 'role', 'login_required');
         }
 
-        this.kuzzle.emit(
-          'core:roleRepository:save',
-          { _id: role._id, controllers: role.controllers });
-
-        return this.persistToDatabase(role, options);
+        this.kuzzle.emit('core:roleRepository:save', {
+          _id: role._id,
+          controllers: role.controllers
+        });
       })
+      .then(() => this.checkRoleNativeRights(role))
+      .then(err => {
+        err = this.checkRolePluginsRights(role);
+        errors = err;
+      })
+      .then(() => this.persistToDatabase(role, options))
       .then(() => this.loadOneFromDatabase(role._id))
       .then(updatedRole => {
         this.roles.set(role._id, updatedRole);
+        if (errors.length) {
+          errors = errorsManager.get('security', 'role', 'invalid_plugin_rights', role._id, errors);
+          errors.role = updatedRole;
+          throw errors;
+        }
         return updatedRole;
       });
   }
 
+  /**
+   * Given a Role object, checks if its controllers and actions exist.
+   *
+   * @param {Role} role
+   */
+  checkRoleNativeRights(role) {
+    const roleRightsError = errorsManager.wrap('security', 'role');
+    Object.keys(role.controllers).forEach(roleController => {
+      if (roleController !== '*' && !this.kuzzle.funnel.controllers.has(roleController)) {
+        if (roleController.split('/').length > 1) {
+          // Meaning this is a plugin controller. So we skip it.
+          return;
+        }
+        roleRightsError.throw('unknown_controller', role._id, roleController);
+      }
+      if (roleController === '*') {
+        this.kuzzle.funnel.controllers.forEach(controller => {
+          Object.keys(role.controllers['*'].actions).forEach(action => {
+            if (action !== '*' && !controller._isAction(action)) {
+              roleRightsError.throw('unknown_controller_action', role._id, action, controller);
+            }
+          });
+        });
+      }
+      else {
+        const controller = this.kuzzle.funnel.controllers.get(roleController);
+        const actions = Object.keys(role.controllers[roleController].actions); 
+        actions.forEach(action => {
+          if (action !== '*' && !controller._isAction(action)) {
+            roleRightsError.throw('unknown_controller_action', role._id, action, roleController);
+          }
+        });
+      }
+    });
+  }
+
+  /**
+   * Given a Role object, checks if its controllers and actions exist in plugins.
+   *
+   * @param {Role} role
+   * @returns Array
+   */
+  checkRolePluginsRights(role) {
+    const roleRightsError = errorsManager.wrap('security', 'role');
+    const errors = [];
+    Object.keys(role.controllers).forEach(roleController => {
+      if (roleController.split('/').length === 1) {
+        // Meaning this is not a plugin controller. So we skip it.
+        return;
+      }
+      const pluginName = roleController.split('/')[0];
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Use destructuring syntax for these assignments from "roleController.split('/')". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS3514)